src/Avalonia.Base/PropertyStore/DirectBindingObserver.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using Avalonia.Data;
+using Avalonia.Threading;
+
+namespace Avalonia.PropertyStore
+{
+    internal class DirectBindingObserver<T> : IObserver<T>,
+        IObserver<BindingValue<T>>,
+        IDisposable
+    {
+        private readonly ValueStore _owner;
+        private IDisposable? _subscription;
+
+        public DirectBindingObserver(ValueStore owner, DirectPropertyBase<T> property)
+        {
+            _owner = owner;
+            Property = property;
+        }
+
+        public DirectPropertyBase<T> Property { get;}
+
+        public void Start(IObservable<T> source)
+        {
+            _subscription = source.Subscribe(this);
+        }
+
+        public void Start(IObservable<BindingValue<T>> source)
+        {
+            _subscription = source.Subscribe(this);
+        }
+
+        public void Dispose()
+        {
+            _subscription?.Dispose();
+            _subscription = null;
+            _owner.OnLocalValueBindingCompleted(Property, this);
+        }
+
+        public void OnCompleted() => _owner.OnLocalValueBindingCompleted(Property, this);
+        public void OnError(Exception error) => OnCompleted();
+
+        public void OnNext(T value)
+        {
+            if (Dispatcher.UIThread.CheckAccess())
+            {
+                _owner.Owner.SetDirectValueUnchecked<T>(Property, value);
+            }
+            else
+            {
+                // To avoid allocating closure in the outer scope we need to capture variables
+                // locally. This allows us to skip most of the allocations when on UI thread.
+                var instance = _owner.Owner;
+                var property = Property;
+                var newValue = value;
+                Dispatcher.UIThread.Post(() => instance.SetDirectValueUnchecked(property, newValue));
```

Comment:
I think we have Post method overload that accepts an object parameter.
Something like:
```
Dispatcher.UIThread.Post(static tuple =>
{
var (instance, property, newValue) = ValueTuple<InstanceType, DirectPropertyBase<T>, T>tuple;
instance.SetDirectValueUnchecked(property, newValue);
}, (_owner.Owner, Property, value));
```
Which will eliminate closures completely.

Also, that's an ugly way to do it, but hey.

Reply:
If this code isn't too ugly, it can be used in multiple places of this PR and later in the codebase in overall.

Reply:
This code is unchanged from the [equivalent of this class on `master`](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaObject.cs#L945-L961). If we decide to do this in a different way, we can do it as a separate PR IMO.

Reply:
Yeah I wrote it quite some time ago before we had parametrized `Post`, but even now that API won't allow for really cheap job posting.